File: apps/web/src/api/graph.ts
import { request } from '@/service/request'

export interface GraphNode {
  name: string
  category: string
  description: string
  connections: number
}

export interface GraphEdge {
  source: string
  target: string
  relation: string
  weight: number
}

export interface GraphData {
  nodes: GraphNode[]
  edges: GraphEdge[]
}

export interface GraphStats {
  concepts: number
  relations: number
}

export const graphApi = {
  getGraph: (limit = 200) =>
    request.get<unknown, GraphData>('/graph', { params: { limit } }),

  getStats: () =>
    request.get<unknown, GraphStats>('/graph/stats'),

  searchConcepts: (q: string, limit = 20) =>
    request.get<unknown, GraphNode[]>('/graph/search', { params: { q, limit } }),

  getNeighbors: (conceptName: string, depth = 2) =>
    request.get<unknown, GraphData>(`/graph/neighbors/${encodeURIComponent(conceptName)}`, {
      params: { depth },
    }),
}

File: apps/api/src/api/routers/graph.py
"""
Knowledge graph API routes.
"""

from fastapi import APIRouter, Query

from ...services.graph import GraphService


def init_router(graph_service: GraphService) -> APIRouter:
    router = APIRouter(prefix="/graph", tags=["graph"])

    @router.get("")
    async def get_graph(limit: int = Query(200, ge=1, le=1000)):
        """Get the full knowledge graph for visualization."""
        return await graph_service.get_graph(limit=limit)

    @router.get("/stats")
    async def get_stats():
        """Get graph statistics (concept count, relation count)."""
        return await graph_service.get_stats()

    @router.get("/search")
    async def search_concepts(
        q: str = Query(..., min_length=1),
        limit: int = Query(20, ge=1, le=100),
    ):
        """Search concepts by name."""
        return await graph_service.search_concepts(query=q, limit=limit)

    @router.get("/neighbors/{concept_name}")
    async def get_neighbors(
        concept_name: str,
        depth: int = Query(2, ge=1, le=5),
    ):
        """Get a concept and its neighbors up to N hops."""
        return await graph_service.get_neighbors(concept_name, depth=depth)

    return router
